### PR TITLE
Proxy changes for vstest task

### PR DESCRIPTION
**Pull request by prawalagarwal**
VSTest task side changes to support proxy for distributed test clients.
1) It removes the agent registration in the distributed flow.
2) It removes the call to TestExecution in the distributed flow.

Both the above calls have been moved to DtaExecutionHost exe call.

### CODE REVIEW

File: Tasks/VsTest/distributedtest.ts [hunk 1]
```diff
@@ -68,25 +50,27 @@ export class DistributedTest {
         }
     }
 
-    private async startDtaExecutionHost(agentId: any) {
-        this.testSourcesFile = this.createTestSourcesFile();
+    private async startDtaExecutionHost() {
+        const dtaExecutionHostTool = tl.tool(path.join(__dirname, 'Modules/DTAExecutionHost.exe'));
         const envVars: { [key: string]: string; } = process.env;
```
Comment (cltshivash) [inline Tasks/VsTest/distributedtest.ts]: nitpick : can we move this closer to the usage ?

File: Tasks/VsTest/distributedtest.ts [hunk 2]
```diff
@@ -222,12 +182,10 @@ export class DistributedTest {
         }
     }
 
-    private async startDtaTestRun() {
-        const runDistributesTestTool = tl.tool(path.join(__dirname, 'modules/TestExecutionHost.exe'));
-        const envVars: { [key: string]: string; } = process.env;
+    private async addDtaTestRunEnvVars(envVars: any) {
         utils.Helper.addToProcessEnvVars(envVars, 'accesstoken', this.dtaTestConfig.dtaEnvironment.patToken);
         utils.Helper.addToProcessEnvVars(envVars, 'environmenturi', this.dtaTestConfig.dtaEnvironment.environmentUri);
```
Comment (cltshivash) [inline Tasks/VsTest/distributedtest.ts]: not needed.. please verify

File: Tasks/VsTest/models.ts [hunk 3]
```diff
@@ -41,6 +41,12 @@ export interface TestConfigurations {
     rerunFailedThreshold: number;
     rerunMaxAttempts: number;
     toolsInstallerConfig: ToolsInstallerConfiguration;
+
+    // Proxy properties
+    proxyUrl: string;
+    proxyUserName: string;
+    proxyPassword: string;
+    proxyBypassHosts: string;
```
Comment (cltshivash) [inline Tasks/VsTest/models.ts]: can this be a separate interface that is contained inside DtaTestConfigurations ?

File: Tasks/VsTest/vstest.ts [hunk 4]
```diff
@@ -433,7 +432,7 @@ async function runVStest(settingsFile: string, vsVersion: number): Promise<tl.Ta
     }
 
     let testselector = new testselectorinvoker.TestSelectorInvoker();
-    let code = testselector.publishCodeChanges(tiaConfig, testCaseFilterFile, vstestConfig.taskInstanceIdentifier);
+    let code = testselector.publishCodeChanges(tiaConfig, vstestConfig, null, testCaseFilterFile, vstestConfig.taskInstanceIdentifier);
```
Comment (cltshivash) [inline Tasks/VsTest/vstest.ts]: >taskInstanceIdentifier [](start = 111, length = 22)

vstestconfig is being passed.. taskInstanceIdentifier is redundant.

File: Tasks/VsTest/testselectorinvoker.ts [hunk 5]
```diff
@@ -7,7 +7,7 @@ import { Helper } from './helpers'
 let perf = require('performance-now');
 
 export class TestSelectorInvoker {
-    public publishCodeChanges(tiaConfig: models.TiaConfiguration, testCaseFilterFile: string, taskInstanceIdentifier: string): number {
+    public publishCodeChanges(tiaConfig: models.TiaConfiguration, vstestConfig: models.VsTestConfigurations, dtaTestConfig: models.DtaTestConfigurations, testCaseFilterFile: string, taskInstanceIdentifier: string): number {
```
Comment (cltshivash) [inline Tasks/VsTest/testselectorinvoker.ts]: >dtaTestConfig [](start = 109, length = 13)

you should pass the proxyInterface  or vstestConfig (dtaconfig inherits vstestconfig right ?)

File: Tasks/VsTest/distributedtest.ts (hunk 1 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: Done

File: Tasks/VsTest/testselectorinvoker.ts (hunk 5 above)
Comment (prawalagarwal) [reply Tasks/VsTest/testselectorinvoker.ts]: Both dtaconfig and vstestconfig extend testconfig. dtaconfig does not inherit vstestconfig

File: Tasks/VsTest/distributedtest.ts [hunk 6]
```diff
@@ -25,12 +25,12 @@ export class DistributedTest {
 
     public runDistributedTest() {
         this.publishCodeChangesIfRequired();
-        this.registerAndConfigureAgent();
+        this.beginDtaExecutionHost();
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: nit: start

File: Tasks/VsTest/distributedtest.ts [hunk 7]
```diff
@@ -25,12 +25,12 @@ export class DistributedTest {
...
     }
 
     private publishCodeChangesIfRequired(): void {
         if (this.dtaTestConfig.tiaConfig.tiaEnabled) {
-            const code = testSelector.publishCodeChanges(this.dtaTestConfig.tiaConfig, null, this.dtaTestConfig.taskInstanceIdentifier);
+            const code = testSelector.publishCodeChanges(this.dtaTestConfig.tiaConfig, null, this.dtaTestConfig, null);
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: why this change is required?

File: Tasks/VsTest/testselectorinvoker.ts [hunk 8]
```diff
@@ -76,11 +77,15 @@ export class TestSelectorInvoker {
                 'filter': pathFilters,
                 'userMapFile': tiaConfig.userMapFile ? tiaConfig.userMapFile : '',
                 'testCaseFilterResponseFile': testCaseFilterFile ? testCaseFilterFile : '',
+                'proxyurl': vstestConfig ? vstestConfig.proxyConfiguration.proxyUrl : dtaTestConfig.proxyConfiguration.proxyUrl,
+                'proxyusername': vstestConfig ? vstestConfig.proxyConfiguration.proxyUserName : dtaTestConfig.proxyConfiguration.proxyUserName,
+                'proxypassword': vstestConfig ? vstestConfig.proxyConfiguration.proxyPassword : dtaTestConfig.proxyConfiguration.proxyPassword,
```
Comment (nigurr) [inline Tasks/VsTest/testselectorinvoker.ts]: why this is not common? repeated 3 times

File: Tasks/VsTest/task.json [hunk 9]
```diff
@@ -17,7 +17,7 @@
     "version": {
         "Major": 2,
         "Minor": 3,
-        "Patch": 8
+        "Patch": 9
```
Comment (nigurr) [inline Tasks/VsTest/task.json]: it's already 9 in master

File: Tasks/VsTest/distributedtest.ts [hunk 10]
```diff
@@ -100,6 +82,17 @@ export class DistributedTest {
...
+        // Set proxy settings to environment if provided
+        if (!utils.Helper.isNullEmptyOrUndefined(this.dtaTestConfig.proxyConfiguration.proxyUrl)) {
+            utils.Helper.addToProcessEnvVars(envVars, 'DTA.ProxyUrl', this.dtaTestConfig.proxyConfiguration.proxyUrl);
+            utils.Helper.addToProcessEnvVars(envVars, 'DTA.ProxyUsername', this.dtaTestConfig.proxyConfiguration.proxyUserName);
+            utils.Helper.addToProcessEnvVars(envVars, 'DTA.ProxyPassword', this.dtaTestConfig.proxyConfiguration.proxyPassword);
+            utils.Helper.addToProcessEnvVars(envVars, 'DTA.ProxyBypassHosts', this.dtaTestConfig.proxyConfiguration.proxyBypassHosts);
+        }
+
+        // Adding Test Execution Host specific variables
+        await this.addDtaTestRunEnvVars(envVars);
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: why this has to be async?

File: Tasks/VsTest/distributedtest.ts [hunk 11]
```diff
@@ -39,27 +39,9 @@ export class DistributedTest {
...
-            try {
-                if (this.dtaPid !== -1) {
-                    tl.debug('Trying to kill the Modules/DTAExecutionHost.exe process with pid :' + this.dtaPid);
-                    process.kill(this.dtaPid);
-                }
-            } catch (error) {
-                tl.warning('Modules/DTAExecutionHost.exe process kill failed, pid: ' + this.dtaPid + ' , error :' + error);
-            }
+            await this.startDtaExecutionHost();
             tl.setResult(tl.TaskResult.Succeeded, 'Task succeeded');
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: if agent registration fails, you are retuning exit code 1. but where we are marking the task as failed?

File: Tasks/VsTest/testselectorinvoker.ts [hunk 12]
```diff
@@ -76,11 +77,15 @@ export class TestSelectorInvoker {
...
+                'proxybypasslist': vstestConfig ? vstestConfig.proxyConfiguration.proxyBypassHosts : dtaTestConfig.proxyConfiguration.proxyBypassHosts,
                 'AGENT_VERSION': tl.getVariable('AGENT.VERSION'),
-                'VsTest_TaskInstanceIdentifier': taskInstanceIdentifier,
-                'VSTS_HTTP_RETRY' : tl.getVariable('VSTS_HTTP_RETRY'),
-                'VSTS_HTTP_TIMEOUT' : tl.getVariable('VSTS_HTTP_TIMEOUT'),
-                'DebugLogging' : this.isDebugEnabled()
+                'VsTest_TaskInstanceIdentifier': vstestConfig ? vstestConfig.taskInstanceIdentifier : dtaTestConfig.taskInstanceIdentifier,
+                'VSTS_HTTP_RETRY': tl.getVariable('VSTS_HTTP_RETRY'),
```
Comment (nigurr) [inline Tasks/VsTest/testselectorinvoker.ts]: keep naming consistency please

File: Tasks/VsTest/distributedtest.ts (hunk 6 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: Done

File: Tasks/VsTest/distributedtest.ts (hunk 7 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: proxy configuration is part part of both vstest config and dtaconfig. Since publish code changes is invoked in both the flows, we need to pass both.

File: Tasks/VsTest/distributedtest.ts (hunk 11 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: Modified the code to address this issue

File: Tasks/VsTest/distributedtest.ts (hunk 10 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: because this method contains one call to updatesettings which is an async funtion due to historical reasons.

File: Tasks/VsTest/distributedtest.ts (hunk 2 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: Removed

File: Tasks/VsTest/testselectorinvoker.ts (hunk 8 above)
Comment (prawalagarwal) [reply Tasks/VsTest/testselectorinvoker.ts]: Its 3 different properties. Didnt get you

File: Tasks/VsTest/distributedtest.ts [hunk 13]
```diff
@@ -39,54 +39,44 @@ export class DistributedTest {
...
 
-    private async startDtaExecutionHost(agentId: any) {
-        this.testSourcesFile = this.createTestSourcesFile();
+    private async startDtaExecutionHost(): Promise<number> {        
         const envVars: { [key: string]: string; } = process.env;
+        this.testSourcesFile = this.createTestSourcesFile();
         tl.debug('Total env vars before setting DTA specific vars is :' + Object.keys(envVars).length);
         utils.Helper.addToProcessEnvVars(envVars, 'DTA.AccessToken', this.dtaTestConfig.dtaEnvironment.patToken);
-        utils.Helper.addToProcessEnvVars(envVars, 'DTA.AgentId', agentId);
+        utils.Helper.addToProcessEnvVars(envVars, 'DTA.AgentId', '0');
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: remove this

File: Tasks/VsTest/distributedtest.ts [hunk 14]
```diff
@@ -116,74 +117,40 @@ export class DistributedTest {
...
-                console.error(line);
-            });
-        });
-
-        proc.on('error', (err) => {
-            this.cleanUpDtaExeHost();
-            throw new Error('Failed to start Modules/DTAExecutionHost.exe.');
-        });
+        const dtaExecutionHostTool = tl.tool(path.join(__dirname, 'Modules/DTAExecutionHost.exe'));
+        var code = await dtaExecutionHostTool.exec(<tr.IExecOptions>{ cwd: path.join(__dirname, 'Modules'), env: envVars });
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: ??

File: Tasks/VsTest/distributedtest.ts [hunk 15]
```diff
@@ -116,74 +117,40 @@ export class DistributedTest {
...
+            runInParallel: this.dtaTestConfig.runInParallel,
+            settingsType: !utils.Helper.isNullOrUndefined(this.dtaTestConfig.settingsFile) ? this.dtaTestConfig.settingsFile.endsWith('.runsettings') ? 'runsettings' : this.dtaTestConfig.settingsFile.endsWith('.testsettings') ? 'testsettings' : 'none' : 'none',
+            task: "VsTestDistributedFlow",
+            testSelection: this.dtaTestConfig.testSelection,
+            tiaEnabled: this.dtaTestConfig.tiaConfig.tiaEnabled,
+            vsTestVersion: this.dtaTestConfig.vsTestVersionDetails.majorVersion + '.' + this.dtaTestConfig.vsTestVersionDetails.minorversion + '.' + this.dtaTestConfig.vsTestVersionDetails.patchNumber
+        };
+
+        if (code !== 0) {
+            tl.debug('Modules/DTAExecutionHost.exe process exited with code ' + code);
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: remove

File: Tasks/VsTest/distributedtest.ts [hunk 16]
```diff
@@ -251,63 +214,59 @@ export class DistributedTest {
...
-            utils.Helper.addToProcessEnvVars(envVars, 'rerunfailedtests', "true");
+        if (this.dtaTestConfig.rerunFailedTests) {            
+            utils.Helper.addToProcessEnvVars(envVars, 'TE.RerunFailedTests', "true");
             tl.debug("Type of rerun: " + this.dtaTestConfig.rerunType);
             if (this.dtaTestConfig.rerunType === 'basedOnTestFailureCount') {
-                utils.Helper.addToProcessEnvVars(envVars, 'rerunfailedtestcasesmaxlimit', this.dtaTestConfig.rerunFailedTestCasesMaxLimit.toString());
+                utils.Helper.addToProcessEnvVars(envVars, 'TE.RerunFailedTestCasesMaxLimit', this.dtaTestConfig.rerunFailedTestCasesMaxLimit.toString());
             } else {
-                utils.Helper.addToProcessEnvVars(envVars, 'rerunfailedthreshold', this.dtaTestConfig.rerunFailedThreshold.toString());
+                utils.Helper.addToProcessEnvVars(envVars, 'TE.RerunFailedThresholdTE', this.dtaTestConfig.rerunFailedThreshold.toString());
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: remove TE

File: Tasks/VsTest/distributedtest.ts [hunk 17]
```diff
@@ -251,63 +214,59 @@ export class DistributedTest {
...
         }
 
-        if (this.dtaTestConfig.rerunFailedTests) {
-            utils.Helper.addToProcessEnvVars(envVars, 'rerunfailedtests', "true");
+        if (this.dtaTestConfig.rerunFailedTests) {            
+            utils.Helper.addToProcessEnvVars(envVars, 'TE.RerunFailedTests', "true");
             tl.debug("Type of rerun: " + this.dtaTestConfig.rerunType);
             if (this.dtaTestConfig.rerunType === 'basedOnTestFailureCount') {
-                utils.Helper.addToProcessEnvVars(envVars, 'rerunfailedtestcasesmaxlimit', this.dtaTestConfig.rerunFailedTestCasesMaxLimit.toString());
+                utils.Helper.addToProcessEnvVars(envVars, 'TE.RerunFailedTestCasesMaxLimit', this.dtaTestConfig.rerunFailedTestCasesMaxLimit.toString());
```
Comment (nigurr) [inline Tasks/VsTest/distributedtest.ts]: >RerunFailedTestCasesMaxLimit [](start = 62, length = 28)

add in mseng pr

File: Tasks/VsTest/distributedtest.ts (hunk 16 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: done

File: Tasks/VsTest/distributedtest.ts (hunk 15 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: done

File: Tasks/VsTest/distributedtest.ts (hunk 13 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: done

File: Tasks/VsTest/distributedtest.ts (hunk 14 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: removed

File: Tasks/VsTest/distributedtest.ts (hunk 17 above)
Comment (prawalagarwal) [reply Tasks/VsTest/distributedtest.ts]: added

File: Tasks/VsTest/testselectorinvoker.ts (hunk 12 above)
Comment (prawalagarwal) [reply Tasks/VsTest/testselectorinvoker.ts]: Discussed offline
